### PR TITLE
Handle pagination for ELB or NLB lookup

### PR DESCRIPTION
**Pull request by sjmiller609**

### CODE REVIEW

**Comment by sjmiller609** [comment]
@ianstanton go ahead and merge it if the CI passes and you are happy with the change. I'll add some review comments that you might find interesting.

File: files/elb_lookup.py [hunk 1]
```diff
@@ -1,60 +1,99 @@
...
 
-def _get_elb_name():
+    vpc_id = os.environ['VPC_ID']
+
+    # Get the first page of results from both
+    # ELBv1 and ELBv2 APIs
     response = client.describe_load_balancers()
     response2 = clientv2.describe_load_balancers()
     load_balancers = response['LoadBalancerDescriptions']
     nlb_load_balancers = response2['LoadBalancers']
```
Comment (sjmiller609) [inline files/elb_lookup.py:18]: first, we send an api request to get a list of load balancers. there are two different APIs for load balancers, so we ask for both of them. the catch that you discovered is that they will only send some, and not all (i.e. pagination)

File: files/elb_lookup.py [hunk 2]
```diff
@@ -1,60 +1,99 @@
...
-            ResourceArns=nlb_load_balancer_arns_in_vpc)
-        tag_descriptions += response2['TagDescriptions']
-
-    # return the name of the ELB corresponding to this cluster
-    for description in tag_descriptions:
+    v1_next_token = response.get('NextMarker')
+    v2_next_token = response2.get('NextMarker')
+
+    # do-while loop, conditional at bottom
+    while True:
```
Comment (sjmiller609) [inline files/elb_lookup.py:24]: we will loop until we are done paging through all the load balancers, or we find the one we are looking for.

File: files/elb_lookup.py [hunk 3]
```diff
@@ -1,60 +1,99 @@
...
+        load_balancer_names_in_vpc = []
+        for lb in load_balancers:
+            if lb['VPCId'] == vpc_id:
+                load_balancer_names_in_vpc.append(lb["LoadBalancerName"])
+
+        # get all the ELBs in this VPC
+        nlb_load_balancer_arns_in_vpc = []
+        for lb in nlb_load_balancers:
+            if lb['VpcId'] == vpc_id:
+                nlb_load_balancer_arns_in_vpc.append(lb["LoadBalancerArn"])
```
Comment (sjmiller609) [inline files/elb_lookup.py:36]: we are only considering LBs in the expected VPC. the key is named differently between the two APIs.

File: files/elb_lookup.py [hunk 4]
```diff
@@ -1,60 +1,99 @@
...
+        # get the tags by LB ARN for the case of ELBv2
+        if len(nlb_load_balancer_arns_in_vpc) > 0:
+            response2 = clientv2.describe_tags(
+                ResourceArns=nlb_load_balancer_arns_in_vpc)
+            for description in response2['TagDescriptions']:
+                arn = description['ResourceArn']
+                for lb in nlb_load_balancers:
+                    if lb['LoadBalancerArn'] == arn:
+                        yield lb['LoadBalancerName'], description
+                        break
```
Comment (sjmiller609) [inline files/elb_lookup.py:54]: out of the LBs in this VPC, we query the API again (on both - if LBs present in both results) for their tags. the tags are what is used to determine if it's the LB we are looking for

File: files/elb_lookup.py [hunk 5]
```diff
@@ -1,60 +1,99 @@
...
+        for lb in nlb_load_balancers:
+            if lb['VpcId'] == vpc_id:
+                nlb_load_balancer_arns_in_vpc.append(lb["LoadBalancerArn"])
+
+        # get the tags by LB name for the case of ELBv1
+        if len(load_balancer_names_in_vpc) > 0:
+            response = client.describe_tags(
+                LoadBalancerNames=load_balancer_names_in_vpc)
+            for description in response['TagDescriptions']:
+                yield description['LoadBalancerName'], description
```
Comment (sjmiller609) [inline files/elb_lookup.py:43]: the "yield" command is used to make python generators - it's like a function that returns something, but execution is handled back to the function after the caller is done. this is python specific terminology. In general, it's known as a coroutine. This is an extremely handy tool for (in this case) for searching through a list of something without querying for everything ahead of time. for example, we will only hold the page size of results in memory until the next page. The alternative is to load all results first, then loop through them, which is less efficient and might run into overly using memory or API calls.

File: files/elb_lookup.py [hunk 6]
```diff
@@ -1,60 +1,99 @@
...
+    ''' Iterate through a generator that returns LB or NLB name and tags description
+    '''
+    # Initialize clients, if necessary
+    if not client:
+        client = boto3.client('elb')
+    if not clientv2:
+        clientv2 = boto3.client('elbv2')
+
+    cluster_name = os.environ['CLUSTER_NAME']
+    for name, description in _generate_lb_name_and_tags(client=client, clientv2=clientv2):
```
Comment (sjmiller609) [inline files/elb_lookup.py:85]: On each loop, the _generate_lb_name_and_tags function is called and the yield is returned into name and description variables. the work is not performed until the start of each loop. that's what makes it a generator (i.e. coroutine)

File: test/test_elb_lookup.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python3
+
+import os
+this_directory = os.path.dirname(os.path.realpath(__file__))
+script_dir = os.path.realpath(os.path.join(this_directory,'../files'))
+import sys
+sys.path.append(script_dir)
+from unittest.mock import MagicMock
```
Comment (sjmiller609) [inline test/test_elb_lookup.py:8]: MagicMock (commonly known as 'mocks') is a useful tool for unit testing when you don't want to actually use the external dependencies. In this case, we want to unit test something that calls out to AWS. Instead, we can provide sample return values using mocks. This allows us to test our logic without actually interacting with AWS.

**Review by ianstanton — APPROVED** [review]
LGTM!

**Comment by ianstanton** [comment]
> @ianstanton go ahead and merge it if the CI passes and you are happy with the change. I'll add some review comments that you might find interesting.

@sjmiller609 I am not authorized to merge.

**Pull request merged**